service/src/BatchServer.cpp

```diff
@@ -0,0 +1,250 @@
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "BatchServer.hpp"
+
+#include <cstdlib>
+#include <cerrno>
+#include <unistd.h>
+
+#include "geopm/Exception.hpp"
+#include "geopm/PlatformIO.hpp"
+#include "geopm/SharedMemory.hpp"
+#include "geopm/Helper.hpp"
+#include "geopm/PlatformIO.hpp"
+#include "BatchStatus.hpp"
+#include "geopm_debug.hpp"
+
+
+namespace geopm
+{
+    std::unique_ptr<BatchServer>
+    BatchServer::make_unique(int client_pid,
+                             const std::vector<geopm_request_s> &signal_config,
+                             const std::vector<geopm_request_s> &control_config)
+    {
+        return geopm::make_unique<BatchServerImp>(client_pid, signal_config,
+                                                  control_config);
+    }
+    BatchServerImp::BatchServerImp(int client_pid,
+                                   const std::vector<geopm_request_s> &signal_config,
+                                   const std::vector<geopm_request_s> &control_config)
+        : BatchServerImp(client_pid, signal_config, control_config,
+                         platform_io(), nullptr, nullptr, nullptr)
+    {
+        // Fork the server when calling real constructor.
+        auto setup = [this]() {this->create_shmem();};
+        auto run = [this]() {this->run_batch();};
+        m_server_pid = fork_with_setup(setup, run);
+    }
+
+    BatchServerImp::BatchServerImp(int client_pid,
+                                   const std::vector<geopm_request_s> &signal_config,
+                                   const std::vector<geopm_request_s> &control_config,
+                                   PlatformIO &pio,
+                                   std::shared_ptr<BatchStatus> batch_status,
+                                   std::shared_ptr<SharedMemory> signal_shmem,
+                                   std::shared_ptr<SharedMemory> control_shmem)
+        : m_client_pid(client_pid)
+        , m_server_key(std::to_string(m_client_pid))
+        , m_signal_config(signal_config)
+        , m_control_config(control_config)
+        , m_pio(pio)
+        , m_signal_shmem(signal_shmem)
+        , m_control_shmem(control_shmem)
+        , m_batch_status(batch_status != nullptr ?
+                         batch_status : BatchStatus::make_unique_server(m_client_pid,
+                                                                        m_server_key))
+        , m_server_pid(0)
+        , m_is_active(true)
+    {
+
+    }
+
+    BatchServerImp::~BatchServerImp()
+    {
+        stop_batch();
```

Comment:
This calls `BatchStatusImp::send_message` and `BatchStatusImp::receive_message`, which may throw exceptions. Consider wrapping this `stop_batch()` call in a `try...except{log the error}` block, to avoid throwing from a destructor (a possible source of undefined-behavior bugs in cases were a destructor is called due to some other exception, then raises a nested exception from itself).

Reply:
Also, is it possible to hang on destruction if it's waiting for a message that never comes?

Reply:
Good idea!